### DISCUSSION TITLE
Can't fetch resources when resources and render targets are in different subdirectories

### DISCUSSION THREAD

**Discussion by kthayashi** [quarto-dev/quarto-cli#9097] [Q&A] [answered]
### Description

I'm finding that `quarto render` is unable to fetch images during render when both the images and render targets are in different subdirectories of a (default) Quarto project (possibly related to #4041 but not sure). I've attached a minimal reproducible example ([quarto-resource-fetch-reprex.zip](https://github.com/quarto-dev/quarto-cli/files/14620425/quarto-resource-fetch-reprex.zip)), which I describe below.

Directory structure:

```
.
├── _quarto.yml
├── images
│   └── img.png
├── postrender.R
├── qmd
│   ├── load-image-v1.qmd
│   ├── load-image-v2.qmd
│   └── make-image.qmd
└── quarto-resource-fetch-reprex.Rproj
```

`_quarto.yml`:

```yml
project: 
  type: default
  execute-dir: project
  render:
    - qmd/make-image.qmd
    - qmd/load-image-v1.qmd
    - qmd/load-image-v2.qmd
  post-render: postrender.R

format:
  html:
    embed-resources: true
```

I'm setting `embed-resources: true` here because I want to highlight behavior that should be compatible with stand-alone output formats like pdf or docx. `execute-dir` needs to be `project`, or else code in the .qmd documents will fail.

`load-image-v1.qmd` runs without interruption upon render (at least for html output; render gets aborted if output is pdf), but returns the warning `[WARNING] Could not fetch resource images/img.png`, and the output document is missing the image:

``````qmd
---
title: Load image (v1)
---

```{r}
#| label: check

getwd()

file.exists("images/img.png")
```

The image cannot be fetched during render:

```{r}
#| label: load-image

knitr::include_graphics("images/img.png")
```
``````

However, temporarily copying the `images` directory from the project root into the `qmd` directory during render makes the image resources available to the render targets, which is what `load-image-v2.qmd` does. The `postrender.R` post-render script is there to remove this temporarily copied directory. Both `load-image-v1.qmd` and `load-image-v2.qmd` can be run without issues in an interactive session in RStudio.

Is this expected behavior of `quarto render`? If so, is there a way to render the targets without having to manually relocate resources while maintaining this sort of directory structure? Or might there be a way (now or in the future) to explicitly specify where resources are to be fetched from during/after render?

I'm running Quarto v1.4.550 on macOS Ventura 13.4. Output of `quarto check` is provided below:

```
Quarto 1.4.550
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.11: OK
      Dart Sass version 1.69.5: OK
      Deno version 1.37.2: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.4.550
      Path: /Applications/quarto/bin

[✓] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Using: Installation From Path
      Path: /Library/TeX/texbin
      Version: 2022

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.11.4
      Path: /opt/homebrew/opt/python@3.11/bin/python3.11
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with python3 -m pip install jupyter

[✓] Checking R installation...........OK
      Version: 4.3.1
      Path: /Library/Frameworks/R.framework/Resources
      LibPaths:
        - /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library
      knitr: 1.44
      rmarkdown: 2.25

[✓] Checking Knitr engine render......OK
```

**Comment by mcanouil** [comment]
Could you share your setup as a Git repository? (only what's needed to reproduce the issue, everything else should be removed)

Why are you using `knitr` to add a figure exactly?

```md
![](/images/img.png)
```

Note that paths are resolved from document:
```r
knitr::include_graphics("images/img.png")
```

means/translates to (thus why don't use markdown directly?)

```md
![](images/img.png)
```

There is no `images/img.png` alongside your document.
The image is `../images/img.png`.

**Reply by kthayashi** [reply]
Here's the reprex on GitHub: https://github.com/kthayashi/quarto-resource-fetch-reprex

I've also added attempts at using Markdown syntax to get the images. As you've noted, it does work with the appropriate relative path from the document.

I was using `knitr::include_graphics()` for two main reasons: (1) Out of habit, as it used to make it easier to control output size (e.g. `out.width: "100%"`) etc. for images (at least with Rmarkdown, not sure if this is true with Quarto). (2) It seems to respect paths relative to the working directory (typically the project root), at least in an interactive session (which is where I spend most of my time writing).

> Note that paths are resolved from document:

This is good to have confirmed, but it's still confusing to me that `getwd()` in the rendered documents returns the project root (presumably because of `execute-dir: project`?), the *code* for `knitr::include_graphics() `runs without error, but the rendered *output* can't access the image.

**Reply by mcanouil** [reply]
`knitr::include_graphics()` only writes the markdown for you. It does nothing else. (It basically is more typing than direct use of markdown syntax, by the way described in Quarto documentation and other markdown related documentation).

`execute-dir` is for computations.

So you have to make sure your code produces valid paths and markdown.

Computations are nearly the first thing done.
Then the intermediate markdown is used and converted to the requested format.
You can use `keep-md: true` to see what you are producing.

Here the issue is that you are producing invalid markdown image path.

**Reply by kthayashi** [reply]
Ok, that makes sense. So is this behavior of paths being resolved from the document specific to resources such as images that somehow need to be referenced relative to the document at the time of render? And I can write/read data (e.g. .csv files) using paths relative to the execution directory as opposed to the document location (e.g. `read.csv("data/df.csv")`) because this is fully contained within computations?

**Reply by mcanouil** [reply, marked as answer]
Whatever is done in the code cells is done based on the working directory which can be set as you did.

Once computations are done, anything that relies on paths should start with `/` for absolute to project root paths which in case of single document is the document, else it's where `_quarto.yml` is.
Or use paths relative to the document calling/using the paths in question.

**Reply by kthayashi** [reply]
Ok, thank you very much - this really helped clarify things for me!
